Pending retries message redirects

boblangley opened the pull request:
Connects to https://github.com/Particular/V6Launch/issues/77

### Improvements
This release adds support for creating failed message redirects to override the source failed message queue in the case of permanent decommissioning. Additionally this release adds support for listing failed messages that have been been returned for retry, but have not yet been seen again by ServiceControl (either failed again or successful via auditing). 

### Should You Upgrade Immediately
Users will only be required to update to support the version of ServicePulse that supports message redirects and pending retries as described above.

boblangley commented:
@WojcikMike I re-combined the 2 PRs into 1 to make it easier. Check the nancy module for message redirects, there are few changes to the status codes (mostly Accepted to OK) and perhaps some metadata (e-tags are no longer relevant).

boblangley commented:
I pushed updates making sure they do not cause dependencies. I also re-based, squashing the earlier changes into a single commit.

WojcikMike commented on src/ServiceControl/MessageFailures/Api/QueueAddressModule.cs:
```diff
@@ -0,0 +1,52 @@
+                    if (failedMessageQueues.Count < take)
+                    {
+                        failedMessageQueues.AddRange(session.Query<QueueAddress, QueueAddressIndex>()
+                            .Where(q => !queueAddresses.Contains(q.PhysicalAddress) && q.PhysicalAddress.Contains(search))
```
This is throwing an exception while trying to call Contains. RavenDB is not supporting that method.

johnsimons commented on src/ServiceControl/MessageFailures/Api/FailedMessageViewIndex.cs:
```diff
@@ -30,6 +31,7 @@ select new
                     message.Status,
                     TimeSent = (DateTime) processingAttemptsLast.MessageMetadata["TimeSent"],
                     ReceivingEndpointName = ((EndpointDetails) processingAttemptsLast.MessageMetadata["ReceivingEndpoint"]).Name,
+                    QueueAddress = processingAttemptsLast.FailureDetails.AddressOfFailingEndpoint.ToLowerInvariant(),
```
@boblangley why are we `ToLowerInvariant()` in an index ?
We should do that as part of the display not the index

boblangley replied:
Will we need it when we filter by QueryAddress?

https://github.com/Particular/ServiceControl/pull/762/files#diff-59f9f660f25fbc5f2a406e858443f133R233

johnsimons commented on src/ServiceControl/MessageRedirects/Api/MessageRedirectsModule.cs:
```diff
@@ -0,0 +1,201 @@
+                    }
+
+                    messageRedirect.ToPhysicalAddress = request.tophysicaladdress.ToLowerInvariant();
+                    messageRedirect.LastModifiedTicks = DateTime.UtcNow.Ticks;
```
Why do we need this ?

boblangley commented:
@Particular/servicecontrol-maintainers @WojcikMike  Please review PR:

To support pending retries listing added failed message queue to filtering.
- Failed messages now include the physical queue address property as `QueueAddress`
- **NEW** GET `/errors/queues/addresses` List failed message queue physical addresses. Supports paging.
- **NEW** GET `/errors/queues/addresses/search/{search}` List failed message queue physical addresses filtered by startswith search term. Supports paging.
- **UPDATED** GET `/errors` Add support for `queueaddress` List failed messages. Added support for query parameter to filter by failed message queue physical address.
- **NEW** POST `/errors/queues/{queueaddress}/retry` Retry failed messages by physical queue address by status. Default status when omitted is `RetryIssued`. `Archived` and `Resolved` status will be rejected.

To support message redirects:
- Redirects are a From and a To. From (toLower) is the unique identifier which is sometimes represented using a deterministic GUID.
- Redirects are stored in a single document. This prevents query index issues. Document Id is `messageredirects`. Initial estimate was ~10k redirects per MB document size.
- Failed message retry will replace failed queue address with redirect To when failed queue address matches from, case insensitive.
- The redirects document is re-loaded for each batch of retries.
- **NEW** POST `/redirects` Create a new redirect. 
   - Validate from/to are populated. 
   - Validate From is unique (ignore if To is the same). 
   - Validate no existing To point to new From.
   - Retry by failed queue address can be called when redirect is created (for pending retries) by including `retryexisting` query parameter
- **NEW** PUT `/redirects/{messageredirectid:guid}/` Update existing redirect. Validate To is populated. Validate From exists. Validate To does not point to any existing From.
- **NEW** HEAD `/redirects` Get stats for redirects.
- **NEW** GET `/redirects` List existing redirects. Supports paging.

johnsimons commented on src/ServiceControl/MessageRedirects/Api/MessageRedirectsModule.cs:
```diff
@@ -0,0 +1,200 @@
+
+                    Bus.Publish(messageRedirectChanged);
+
+                    return HttpStatusCode.OK;
```
I wonder if we should return a 204 ?

boblangley replied:
As you stated in slack both are acceptable, but I think 204 is the more appropriate.

johnsimons commented on src/ServiceControl/MessageRedirects/Api/MessageRedirectsModule.cs:
```diff
@@ -0,0 +1,200 @@
+                    });
+                }
+
+                return HttpStatusCode.OK;
```
I wonder if we should return a 204 ?

boblangley replied:
As you stated in slack both are acceptable, but I think 204 is the more appropriate.

WojcikMike commented on src/ServiceControl/MessageRedirects/Api/MessageRedirectsModule.cs:
```diff
@@ -0,0 +1,202 @@
+
+                    return Negotiate
+                        .WithModel(queryResult)
+                        .WithEtagAndLastModified(redirects.ETag, queryResult.Max(r => r.LastModified))
```
When the collection is empty the Max throws Invalid Operation Expcetion:  Sequence contains no elements

WojcikMike commented on src/ServiceControl/MessageRedirects/Api/MessageRedirectsModule.cs:
```diff
@@ -0,0 +1,202 @@
+                }
+            };
+
+            Get["/redirects"] = _ =>
```
I have a problem with that call. I have 2 redirects, I remove 1 using delete. Now when I call this method again, even that when I debug in code I see that there is only 1 redirect, after return browser shows me that response has a status 304 and 2 redirects. Something is off with Nancy recognizing that results are not changed

boblangley replied:
Addressed in https://github.com/Particular/ServiceControl/pull/762/commits/9cbcb86519b7107bf92a5e7daf2fdb6f3f6c144c Used document last modified metadata instead of internal data which may not be complete or change when the document is updated

boblangley commented:
Fixed issue with `queue_address` not being populated with https://github.com/Particular/ServiceControl/pull/762/commits/e10a0bee98935d73c88d29284f634f2855662fd9. Updated comment https://github.com/Particular/ServiceControl/pull/762#issuecomment-234110999 to reflect the correct name of the property when calling the api

boblangley commented:
@WojcikMike I updated Message Redirects PUT/DELETE to return 204 on success instead of 200.

boblangley commented:
@Particular/servicecontrol-maintainers @WojcikMike SC is ready. ServicePulse will require some adjustments. The Pending Retries actions have been changed (parameters moved from URI to POST body) and we need to test long queue names (ala Azure Storage Queues) with the ServicePulse UI.

boblangley commented:
Merged #762
